### PR TITLE
Implementation of discord webhook

### LINKED ISSUE #2191: Change webhook format for Discord compatibility
- Gitea version (or commit ref):
1.1.2
- Can you reproduce the bug at https://try.gitea.io:
  - [x] No (It's currently down)

## Description

Currently webhook uses the format `repo:[branch]`
This format parsed incorrectly in Discord but if it were to be changed from `[ ]` to `( )` then this would be fixed. 

This is already fixed in gogs ([gogits/gogs/issues/4088](https://github.com/gogits/gogs/issues/4088)). I hope it can be done in gitea. 

## Screenshots

screenshot 134

### PR DESCRIPTION
**Pull request by lunny**
will fix #2191

### CODE REVIEW

File: models/webhook_discord.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"code.gitea.io/git"
+	api "code.gitea.io/sdk/gitea"
+
```
Comment (appleboy) [inline models/webhook_discord.go]: remove this empty line.

Comment (lunny) [reply models/webhook_discord.go]: above is other packages on gitea, below is submodule on gitea repo.

Comment (appleboy) [reply models/webhook_discord.go]: Got it 👍

File: models/webhook_discord.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		if s, err := strconv.ParseInt(clr, 16, 32); err == nil {
+			return int(s)
+		}
+	}
+
+	return 0
+}
+
+var (
+	successColor = color("#1ac600")
```
Comment (tboerger) [inline models/webhook_discord.go]: Why do you add the # when it's anyway removed within the color function?

File: models/webhook_discord.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	warnColor    = color("#ffd930")
+	failedColor  = color("#ff3232")
+)
+
+// SetSecret sets the slack secret
```
Comment (tboerger) [inline models/webhook_discord.go]: Slack?

File: models/webhook_discord.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+func (p *DiscordPayload) SetSecret(_ string) {}
+
+// JSONPayload Marshals the SlackPayload to json
+func (p *DiscordPayload) JSONPayload() ([]byte, error) {
```
Comment (tboerger) [inline models/webhook_discord.go]: Slack?

File: models/webhook_discord.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+					Name:    p.Sender.UserName,
+					URL:     setting.AppURL + p.Sender.UserName,
+					IconURL: p.Sender.AvatarURL,
+				},
+			},
+		},
+	}, nil
+}
+
+// GetDiscordPayload converts a slack webhook into a SlackPayload
```
Comment (tboerger) [inline models/webhook_discord.go]: Slack?

File: models/webhook_discord.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+func GetDiscordPayload(p api.Payloader, event HookEventType, meta string) (*DiscordPayload, error) {
+	s := new(DiscordPayload)
+
+	discord := &DiscordMeta{}
+	if err := json.Unmarshal([]byte(meta), &discord); err != nil {
+		return s, errors.New("GetSlackPayload meta json:" + err.Error())
```
Comment (tboerger) [inline models/webhook_discord.go]: Slack?

File: modules/auth/repo_form.go [hunk 7]
```diff
@@ -183,6 +183,19 @@ func (f *NewSlackHookForm) Validate(ctx *macaron.Context, errs binding.Errors) b
 	return validate(errs, ctx.Data, f, ctx.Locale)
 }
 
+// NewDiscordHookForm form for creating slack hook
```
Comment (tboerger) [inline modules/auth/repo_form.go]: Slack?

File: routers/repo/webhook.go [hunk 8]
```diff
@@ -443,6 +503,48 @@ func SlackHooksEditPost(ctx *context.Context, form auth.NewSlackHookForm) {
 	ctx.Redirect(fmt.Sprintf("%s/settings/hooks/%d", orCtx.Link, w.ID))
 }
 
+// DiscordHooksEditPost response for editing slack hook
```
Comment (tboerger) [inline routers/repo/webhook.go]: Slack?

File: templates/repo/settings/hook_discord.tmpl [hunk 9]
```diff
@@ -0,0 +1,19 @@
+{{if eq .HookType "discord"}}
+	<p>{{.i18n.Tr "repo.settings.add_discord_hook_desc" "https://discordapp.com" | Str2html}}</p>
+	<form class="ui form" action="{{.BaseLink}}/settings/hooks/discord/{{if .PageIsSettingsHooksNew}}new{{else}}{{.Webhook.ID}}{{end}}" method="post">
+		{{.CsrfTokenHtml}}
+		<div class="required field {{if .Err_PayloadURL}}error{{end}}">
+			<label for="payload_url">{{.i18n.Tr "repo.settings.payload_url"}}</label>
+			<input id="payload_url" name="payload_url" type="url" value="{{.Webhook.URL}}" autofocus required>
+		</div>
+		<div class="field">
+			<label for="username">{{.i18n.Tr "repo.settings.slack_username"}}</label>
```
Comment (tboerger) [inline templates/repo/settings/hook_discord.tmpl]: Slack?

File: templates/repo/settings/hook_discord.tmpl [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+			<input id="username" name="username" value="{{.DiscordHook.Username}}" placeholder="e.g. Gitea">
+		</div>
+		<div class="field">
+			<label for="icon_url">{{.i18n.Tr "repo.settings.slack_icon_url"}}</label>
+			<input id="icon_url" name="icon_url" value="{{.DiscordHook.IconURL}}" placeholder="e.g. https://example.com/img/favicon.png">
```
Comment (tboerger) [inline templates/repo/settings/hook_discord.tmpl]: Slack?

**Comment by lunny** [comment]
@tboerger done.

**Review by tboerger — APPROVED** [review]

**Comment by tboerger** [comment]
LGTM

File: models/webhook_discord.go (hunk 1 above)
Comment (bkcsoft) [reply models/webhook_discord.go]: In that case they should be in the other order.

```
builtin
local package
external package
```

Comment (lunny) [reply models/webhook_discord.go]: @bkcsoft done.

**Comment by lunny** [comment]
rebased.

**Comment by bkcsoft** [comment]
LGTM

**Comment by bkcsoft** [comment]
@lunny (or any merger) Please squash the commits into 1 commit

**Pull request merged**